### PR TITLE
nixos/stargazer: init

### PR DESCRIPTION
**Pull request by gaykitty**
###### Description of changes

Add nixos module for stargazer gemini server.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by gaykitty** [comment]
I took my best shot at adding a module. Hope I did good! ^w^

File: nixos/doc/manual/release-notes/rl-2305.section.md [hunk 1]
```diff
@@ -89,6 +89,8 @@ In addition to numerous new and upgraded packages, this release has the followin
 
 - [jellyseerr](https://github.com/Fallenbagel/jellyseerr), a web-based requests manager for Jellyfin, forked from Overseerr. Available as [services.jellyseerr](#opt-services.jellyseerr.enable).
 
+- [stargazer](https://sr.ht/~zethra/stargazer/), a fast and easy to use Gemini server.Available as [services.stargazer](#opt-services.stargazer.enable).
```
Comment (AndersonTorres) [inline nixos/doc/manual/release-notes/rl-2305.section.md:92]:
```suggestion
- [stargazer](https://sr.ht/~zethra/stargazer/), a fast and easy to use Gemini server. Available as [services.stargazer](#opt-services.stargazer.enable).
```

File: nixos/modules/services/web-servers/stargazer.nix [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+    log-ip-partial = ${boolToString cfg.ipLogPartial}
+    request-timeout = ${toString cfg.requestTimeout}
+    response-timeout = ${toString cfg.responseTimeout}
+
+    [:tls]
+    store = ${toString cfg.store}
+    organization = ${cfg.certOrg}
+    gen-certs = ${boolToString cfg.genCerts}
+    regen-certs = ${boolToString cfg.regenCerts}
+    ${if cfg.certLifetime == "0" then "" else cfg.certLifetime}
```
Comment (AndersonTorres) [inline nixos/modules/services/web-servers/stargazer.nix:21]: Shouldn't this be a number?

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: nixos/modules/services/web-servers/stargazer.nix (hunk 2 above)
Comment (gaykitty) [reply nixos/modules/services/web-servers/stargazer.nix:21]: No, the value is supposed to be an amount of time like 3y for 3 years. There is always a letter at the end. The only way to have an infinate lifetime is to not include that parameter at all. Is there a better way to do an optional parameter?

Comment (AndersonTorres) [reply nixos/modules/services/web-servers/stargazer.nix:21]: Oh, the time parameter is written as a string? Curious.

Nonetheless, I believe you should use optionalString:

 ```suggestion
    ${optionalString (cfg.certLifetime != "0") cfg.certLifetime}
```

P.S.: shouldn't this be something like 

 ```suggestion
    cert-lifetime = ${optionalString (cfg.certLifetime != "0") cfg.certLifetime}
```

?

Comment (gaykitty) [reply nixos/modules/services/web-servers/stargazer.nix:21]: "0" is just a placeholder value that tells it not to put that option in the config file.

Comment (gaykitty) [reply nixos/modules/services/web-servers/stargazer.nix:21]: Ah yes, my mistake.

**Review by felschr — CHANGES_REQUESTED** [review]

File: nixos/modules/services/web-servers/stargazer.nix [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+    organization = ${cfg.certOrg}
+    gen-certs = ${boolToString cfg.genCerts}
+    regen-certs = ${boolToString cfg.regenCerts}
+    ${optionalString (cfg.certLifetime != "") "cert-lifetime = ${cfg.certLifetime}"}
+
+  '';
+  routesFile = routesFormat.generate "router.ini" cfg.routes;
+  configFile = pkgs.runCommand "config.ini" { } ''
+    cat ${globalFile} ${routesFile} > $out
+  '';
```
[the remark this answers is not in the document]
Comment (gaykitty) [reply nixos/modules/services/web-servers/stargazer.nix:27]: I believe that's not currently possible for this package due to #107491 and stargazer requiring global ini values.

**Review by AndersonTorres — APPROVED** [review]
Ping me 3 days to merge.

**Comment by gaykitty** [comment]
> Ping me 3 days to merge.

Are you asking me to ping you at some point or is this for like a bot or something?

**Comment by AndersonTorres** [comment]
> > Ping me 3 days to merge.
> 
> Are you asking me to ping you at some point or is this for like a bot or something?

No, this is not a directive to any bot.

Sometimes I like to give a small grace period, in order to
- someone that notices any issue can give additional advice
- not to keep the PR stuck in a "approved-but-not-merged" state.

If no one raises objections, after 3 days, then please ping me and I will manually merge this.

If anyone raises objections, then please solve them and ping me to a final review.

**Comment by gaykitty** [comment]
Okay sounds good 👍

**Comment by gaykitty** [comment]
@AndersonTorres it's been three days, nothing new

**Comment by gaykitty** [comment]
Rats. Of course once you merge it I find an issue. The order of the sections in the ini file matters with stargazer. So the order they're listed in in the user's confiug needs to be preserved. @AndersonTorres any ideas? I assume that means routes needs to be a list instead of a set.

**Comment by gaykitty** [comment]
I might just update stargazer itself to not behave that way since it's kinda confusing.

**Comment by AndersonTorres** [comment]
Try to fix this and send a PR.

**Pull request merged**